### ISSUE TITLE
EOTS Manager Health Check

### ISSUE THREAD

**Issue by bap2pecs** [babylonchain/finality-provider#487]
## Problem

in #439, I noted that it's very easy to meet with "the EOTS manager server is not responding" error, thus making the e2e tests flaky.

```
        	Error Trace:	/Users/zidong/Documents/Projects/babylon-finality-provider/itest/opstackl2/op_test_manager.go:170
        	            				/Users/zidong/Documents/Projects/babylon-finality-provider/itest/opstackl2/op_e2e_test.go:119
        	Error:      	Received unexpected error:
        	            	the EOTS manager server is not responding: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:12582: connect: connection refused"
        	Test:       	TestOpMultipleFinalityProviders
```

the root cause is in this function of eotsmanager/client/rpcclient.go:

```
func NewEOTSManagerGRpcClient(remoteAddr string) (*EOTSManagerGRpcClient, error) {
	conn, err := grpc.NewClient(remoteAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("failed to build gRPC connection to %s: %w", remoteAddr, err)
	}

	gClient := &EOTSManagerGRpcClient{
		client: proto.NewEOTSManagerClient(conn),
		conn:   conn,
	}

	if err := gClient.Ping(); err != nil {
		return nil, fmt.Errorf("the EOTS manager server is not responding: %w", err)
	}

	return gClient, nil
}
```

so it's likely that the gRPC connection created by `grpc.NewClient()` is not ready yet. and `Ping()` is a gRPC call defined in `EOTSManagerClient`

thus, there is a race condition and a chance to fail

## Proposal

in `google.golang.org/grpc@v1.64.0/clientconn.go`, there is 

```
// GetState returns the connectivity.State of ClientConn.
//
// # Experimental
//
// Notice: This API is EXPERIMENTAL and may be changed or removed in a later
// release.
func (cc *ClientConn) GetState() connectivity.State {
	return cc.csMgr.getState()
}
```

we might use but this API is experimental.

but I think `Ping()` is not the standard way to do health check in gRPC. we should do sth like https://github.com/grpc/grpc-go/tree/master/examples/features/health (but this is also a gRPC call so we might still have to use `GetState()`)

cc @SebastianElvis @gitferry what's your thought?

**Comment by lesterli** [comment]
An example of wait for ready from grpc repo: https://github.com/grpc/grpc-go/blob/master/examples/features/wait_for_ready/main.go

**Comment by gitferry** [comment]
The `eotsd` needs to be started before the finality provider instance as we have a naive implementation of `ping` [here](https://github.com/babylonchain/finality-provider/blob/37429abec0a514c4dbf95074fb231ff8464fdca8/eotsmanager/client/rpcclient.go#L42). Could the error be due to any cases where the finality provider instance is started without the `eotsd` running?

**Comment by bap2pecs** [comment]
> Could the error be due to any cases where the finality provider instance is started without the `eotsd` running?

unlikely. the error was thrown here:

```
	// prepare EOTS manager
	eotsHomeDir := filepath.Join(testDir, "eots-home")
	eotsCfg := eotsconfig.DefaultConfigWithHomePath(eotsHomeDir)
	eh := e2eutils.NewEOTSServerHandler(t, eotsCfg, eotsHomeDir)
	eh.Start()
	eotsCli, err := client.NewEOTSManagerGRpcClient(cfg.EOTSManagerAddress)
	require.NoError(t, err)
```

it throws when we are staring the EOTS manager calling `NewEOTSManagerGRpcClient()`. so I don't think it's related to FP instance

**Comment by gitferry** [comment]
I see. It also makes sense because within `NewEOTSManagerGRpcClient` we would call `client.Ping`. I think a quick fix would be to introduce timeout to the context when calling [Ping](https://github.com/babylonchain/finality-provider/blob/74a575a4840a801644bfb5b7b2473b88cf31c56c/eotsmanager/client/rpcclient.go#L45)

**Comment by bap2pecs** [comment]
> I see. It also makes sense because within `NewEOTSManagerGRpcClient` we would call `client.Ping`. I think a quick fix would be to introduce timeout to the context when calling [Ping](https://github.com/babylonchain/finality-provider/blob/74a575a4840a801644bfb5b7b2473b88cf31c56c/eotsmanager/client/rpcclient.go#L45). We have a reference implementation [here](https://github.com/babylonchain/staking-queue-client/pull/25/files)

thinking more about this. adding a timeout likely won't fix it b/c the current code uses `context.Background()` which doesn't even have a timeout.

I just sent out a fix #517 to add `waitForConnReady` as a quick fix

**Comment by bap2pecs** [comment]
I believe the root cause is now all fixed in https://github.com/babylonchain/finality-provider/commit/ced8564e6b49ba91b4189144592d34631bd364c2
